starknet_devnet/blueprints/rpc/schema.py

```diff
@@ -0,0 +1,237 @@
+                    // ...
+                }
+            }
+        },
+        "param2: {
+            // schema 2
+            "components": {
+                "schemas": {
+                    // rest of the schemas
+                    // ...
+                }
+            }
+        },
+        // ...
+    }
+
+    See _response_schema_for_method docstring for more detailed explanation.
+    """
+    methods, schemas = _load_schemas()
+    params_json: List[Dict[str, Any]] = methods[name]["params"]
+
+    request_schemas = OrderedDict()
+    for param in params_json:
+        name = param["name"]
+        is_required = param.get("required", False)
+        schema = {
+            **param["schema"],
+            "is_required": is_required,
+            "components": {"schemas": schemas},
+        }
+        request_schemas[name] = schema
+
+    return request_schemas
+
+
+def _assert_valid_rpc_schema(data: Dict[str, Any], method_name: str):
+    """
+    Check if rpc response is valid against the schema for given method name
+    """
+    schema = _response_schema_for_method("starknet_" + method_name)
+    validate(data, schema=schema)
+
+
+def _assert_valid_rpc_request(*args, method_name: str, **kwargs):
+    """
+    Validate if RPC request (parameters) is correct.
+
+    Raise ValidationError if not.
+    """
+    schemas = _request_schemas_for_method("starknet_" + method_name)
+
+    if args and kwargs:
+        raise ValueError("Cannot validate schemas with both args and kwargs provided.")
+
+    if args:
+        if len(args) > len(schemas):
+            raise ValidationError("Too many arguments provided.")
+
+        for name, arg in zip_longest(schemas.keys(), args, fillvalue="missing"):
+            if arg == "missing":
+                # fmt: off
+                raise ValidationError(f"Missing positional argument \"{name}\".")
+                # fmt: on
+
+            validate(arg, schemas[name])
+        return
+
+    if kwargs:
+        if len(kwargs) > len(schemas):
+            raise ValidationError("Too many arguments provided.")
+
+        for name, schema in schemas.items():
+            if name not in kwargs:
+                if schema["is_required"]:
+                    # fmt: off
+                    raise ValidationError(f"Missing keyword argument \"{name}\".")
+                    # fmt: on
+                continue
+
+            value = kwargs[name]
+            validate(value, schema)
+        return
+
+    if length := len(schemas) != 0:
+        raise ValidationError(
+            f"0 arguments provided to function expecting {length} arguments."
+        )
+
+
+class ParamsValidationErrorWrapper(Exception):
+    """
+    Wrapper for ValidationError raised during request validation
+    """
+
+    def __init__(self, err: ValidationError):
+        super().__init__("Failed to validate schema for params.")
+        self.validation_error = err
+
+    def __str__(self):
+        # fmt: off
```

Comment:
Here and in other places, instead of toggling formatting off and on, we can simply rely on using triple quotes:
```python
f"""Got invalid value for parameter: "{self.validation_error.message}""""
```